src/charm.py

```diff
@@ -1,19 +1,479 @@
+        """
+        unit_id = unit.name.split("/")[1]
+        return f"{self.app.name}-{unit_id}.{self.app.name}-endpoints"
+
+    # END: helper functions
+
+    # BEGIN: properties
+    @property
+    def mongos_initialised(self) -> bool:
+        """Check if mongos is initialised."""
+        return "mongos_initialised" in self.app_peer_data
+
+    @mongos_initialised.setter
+    def mongos_initialised(self, value: bool):
+        """Set the mongos_initialised flag."""
+        if value:
+            self.app_peer_data["mongos_initialised"] = str(value)
+        elif "mongos_initialised" in self.app_peer_data:
+            del self.app_peer_data["mongos_initialised"]
+
+    @property
+    def _unit_ip(self) -> str:
+        """Returns the ip address of the unit."""
+        return str(self.model.get_binding(Config.Relations.PEERS).network.bind_address)
+
+    @property
+    def is_external_client(self) -> Optional[str]:
+        """Returns the database requested by the hosting application of the subordinate charm."""
+        if EXTERNAL_CONNECTIVITY_TAG not in self.app_peer_data:
+            return False
+
+        return json.loads(self.app_peer_data.get(EXTERNAL_CONNECTIVITY_TAG))
+
+    @property
+    def database(self) -> Optional[str]:
+        """Returns the database requested by the hosting application of the subordinate charm."""
+        if not self._peers:
+            logger.info("Peer relation not joined yet.")
+            # TODO future PR implement relation interface between host application mongos and use
+            # host application name in generation of db name.
+            return "mongos-database"
+
+        return self.app_peer_data.get(DATABASE_TAG, "mongos-database")
+
+    @property
+    def extra_user_roles(self) -> Set[str]:
+        """Returns the user roles requested by the hosting application of the subordinate charm."""
+        if not self._peers:
+            logger.info("Peer relation not joined yet.")
+            return None
+
+        return self.app_peer_data.get(USER_ROLES_TAG, "default")
+
+    @property
+    def mongos_config(self) -> MongosConfiguration:
+        """Generates a MongoDBConfiguration object for mongos in the deployment of MongoDB."""
+        hosts = [self.get_mongos_host()]
+        port = Config.MONGOS_PORT if self.is_external_client else None
+        external_ca, _ = self.tls.get_tls_files(internal=False)
+        internal_ca, _ = self.tls.get_tls_files(internal=True)
+
+        return MongosConfiguration(
+            database=self.database,
+            username=self.get_secret(APP_SCOPE, Config.Secrets.USERNAME),
+            password=self.get_secret(APP_SCOPE, Config.Secrets.PASSWORD),
+            hosts=hosts,
+            port=port,
+            roles=self.extra_user_roles,
+            tls_external=external_ca is not None,
+            tls_internal=internal_ca is not None,
+        )
+
+    @property
+    def _peers(self) -> Optional[Relation]:
```

Comment:
This is python 3.10 standard:
```suggestion
    def _peers(self) -> Relation | None:
```